### PR TITLE
Fix Twitch provider for the Helix API

### PR DESCRIPTION
**Pull request by zAfLu**
See #242

### CODE REVIEW

**Comment by kinosang** [comment]
LGTM

File: src/AspNet.Security.OAuth.Twitch/TwitchAuthenticationOptions.cs [hunk 1]
```diff
@@ -25,10 +25,28 @@ public TwitchAuthenticationOptions()
...
+            });
+
+            ClaimActions.MapCustomJson(ClaimTypes.Name, user =>
+            {
+                return user["data"]?[0]?["login"]?.ToObject<string>();
+            });
+
+            ClaimActions.MapCustomJson(ClaimTypes.GivenName, user =>
+            {
+                return user["data"]?[0]?["display_name"]?.ToObject<string>();
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Twitch/TwitchAuthenticationOptions.cs]: Is the display name guaranteed to be the first/given name? If no, a different claim name should be used.

Comment (zAfLu) [reply src/AspNet.Security.OAuth.Twitch/TwitchAuthenticationOptions.cs]: We discussed this too but we can not find a claim which really fits. 
```json
{
    "data": [
        {
            "id": "1111",
            "login": "zaflu",
            "display_name": "zAfLu",
            "type": "",
            "broadcaster_type": "partner",
            "description": "",
            "profile_image_url": "LINKTOIMAGE",
            "offline_image_url": "LINKTOIMAE",
            "view_count": 000000,
            "email": "xxxx@yyyy.yy"
        }
    ]
}
```
So the display_name shows the real write shape. 

What claim would you choose?

Comment (kevinchalet) [reply src/AspNet.Security.OAuth.Twitch/TwitchAuthenticationOptions.cs]: Feel free to use a custom claim name. E.g https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.Bitbucket/BitbucketAuthenticationConstants.cs#L16

File: src/AspNet.Security.OAuth.Twitch/TwitchAuthenticationConstants.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    public static class TwitchAuthenticationConstants
+    {
+        public static class Claims
+        {
+            public const string DisplayName = "urn:twitch:displayname";
+            public const string Type = "urn:twitch:type";
+            public const string BroadcasterType = "urn:twitch:broadcastertype";
+            public const string Description = "urn:twitch:description";
+            public const string ProfileImageURL = "urn:twitch:profileimageurl";
+            public const string OfflineImageURL = "urn:twitch:offlineimageurl";
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Twitch/TwitchAuthenticationConstants.cs]: `URL` -> `Url` please.

File: src/AspNet.Security.OAuth.Twitch/TwitchAuthenticationOptions.cs [hunk 3]
```diff
@@ -25,10 +26,52 @@ public TwitchAuthenticationOptions()
...
             UserInformationEndpoint = TwitchAuthenticationDefaults.UserInformationEndpoint;
 
-            Scope.Add("user_read");
+            Scope.Add("user:read:email");
 
-            ClaimActions.MapJsonKey(ClaimTypes.NameIdentifier, "_id");
-            ClaimActions.MapJsonKey(ClaimTypes.Name, "name");
+            ClaimActions.MapCustomJson(ClaimTypes.NameIdentifier, user =>
+            {
+                return user["data"]?[0]?["id"]?.ToObject<string>();
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Twitch/TwitchAuthenticationOptions.cs]: Consider using `.Value<string>("id")` instead of `ToObject()`.

**Comment by kevinchalet** [comment]
@zAfLu when you're done, please also update the 1.0 PR so all the claims are consistent between the two versions.

**Comment by zAfLu** [comment]
both updated :)

**Comment by kevinchalet** [comment]
Merged, thanks for your contribution :+1:

**Pull request closed without merging**